test/test_image.py

```diff
@@ -33,14 +34,44 @@ def get_images(directory, img_ext):
                 yield os.path.join(root, fl)
 
 
+def pil_read_image(img_path):
+    with Image.open(img_path) as img:
+        return torch.from_numpy(np.array(img))
+
+
+def normalize_dimensions(img_pil):
+    if len(img_pil.shape) == 3:
+        img_pil = img_pil.permute(2, 0, 1)
+    else:
+        img_pil = img_pil.unsqueeze(0)
+    return img_pil
+
+
 class ImageTester(unittest.TestCase):
     def test_decode_jpeg(self):
+        conversion = [(None, 0), ("L", 1), ("RGB", 3)]
         for img_path in get_images(IMAGE_ROOT, ".jpg"):
-            img_pil = torch.load(img_path.replace('jpg', 'pth'))
-            img_pil = img_pil.permute(2, 0, 1)
-            data = read_file(img_path)
-            img_ljpeg = decode_jpeg(data)
-            self.assertTrue(img_ljpeg.equal(img_pil))
+            for pil_mode, components in conversion:
+                with Image.open(img_path) as img:
+                    is_cmyk = img.mode == "CMYK"
+                    if pil_mode is not None:
+                        if is_cmyk:
+                            # libjpeg does not support the conversion
+                            continue
+                        img = img.convert(pil_mode)
+                    img_pil = torch.from_numpy(np.array(img))
+                    if is_cmyk:
+                        # flip the colors to match libjpeg
+                        img_pil = 255 - img_pil
+
+                img_pil = normalize_dimensions(img_pil)
+                data = read_file(img_path)
+                img_ljpeg = decode_jpeg(data, components=components)
+
+                # Permit a small variation on pixel values to account for implementation
+                # differences between Pillow and LibJPEG.
+                abs_mean_diff = (img_ljpeg.type(torch.float32) - img_pil).abs().mean().item()
+                self.assertTrue(abs_mean_diff < 2)
```

Comment:
Pillow seems to use different weights than libjpeg to convert RGB to grayscale. The differences are slightly visible but nothing significant. Moreover it seems that on windows, the values don't much 100% either. This is why here I permit for some variation across images.

Reply:
> Moreover it seems that on windows, the values don't much 100% either

This is probably the reason why we were using a saved `.pth` for this file. @andfoy might remember a bit more, but I believe that was the reason.

> Pillow seems to use different weights than libjpeg to convert RGB to grayscale

I assume our conversion to RGB in `transforms.functional.convert_rgb_to_grayscale` matches the output that we have in our implementation for grayscale conversion?

Reply:
>I assume our conversion to RGB in transforms.functional.convert_rgb_to_grayscale matches the output that we have in our implementation for grayscale conversion?

Precisely we use the same weights:
`L = R * 0.2989 + G * 0.5870 + B * 0.1140`

Reply:
Same comment as before. I'm resolving this, but any change to torchvision that changes the pixel values returned by it is a BC-breaking change, and we can't detect it now.